src/lc3_vm/virtual_machine.rs

```diff
@@ -393,7 +433,122 @@ impl VM {
+                            print!("{}", c);
+                            // Increment the memory address
+                            char_address = char_address.wrapping_add(1);
+                        }
+                        io::stdout()
+                            .flush()
+                            .map_err(|err| VMError::Execute(format!("TRAP PUTS: {}", err)))?;
+                    }
+                    Trap::In => {
+                        // Prompt the user for a char
+                        print!("Enter a character: ");
+                        let mut char_buffer: [u8; 1] = [0; 1];
+                        // Read char into buffer
+                        std::io::stdin()
+                            .read_exact(&mut char_buffer)
+                            .map_err(|err| VMError::Execute(format!("TRAP IN: {}", err)))?;
+
+                        let c_u8 = char_buffer.first().ok_or(VMError::Execute(String::from(
+                            "TRAP IN: failed to read char",
+                        )))?;
+                        let c_char: char = (*c_u8).into();
+                        // Echo the character
+                        print!("{c_char}");
+                        // Save char into R0
+                        self.update_register(0, (*c_u8).into())
+                            .map_err(|err| VMError::Execute(format!("TRAP IN: {}", err)))?;
+
+                        self.update_flags(0)
+                            .map_err(|err| VMError::Execute(format!("TRAP IN: {}", err)))?;
+                        io::stdout()
+                            .flush()
+                            .map_err(|err| VMError::Execute(format!("TRAP IN: {}", err)))?;
+                    }
+                    Trap::Putsp => {
+                        // Get starting address of first two chars
+                        let mut char_address = self
+                            .get_register_value(0)
+                            .map_err(|err| VMError::Execute(format!("TRAP PUTSP: {}", err)))?;
+                        while let Ok(Some(c)) = self.read_word(char_address) {
+                            // The string ends when the read word is 0x0000
+                            if c == 0x0000 {
+                                break;
+                            }
+                            // Get the first char
+                            let c1: u8 = ((c & 0b_1111_1111_0000_0000) >> 8)
+                                .try_into()
+                                .map_err(|err| VMError::Execute(format!("TRAP PUSTP: {}", err)))?;
+                            // Get the second char
+                            let c2: u8 = (c & 0b_0000_0000_1111_1111)
+                                .try_into()
+                                .map_err(|err| VMError::Execute(format!("TRAP PUSTP: {}", err)))?;
+
+                            print!("{}", c1);
+                            print!("{}", c2);
```

Comment:
Quick question, why not do:

```rust
                            print!("{} {}", c1, c2);
```

?

Asking simply out of curiosity

Reply:
This should be possible